Genomic range buckets

orodeh opened the pull request:
Mike, please take a look and see if this is any good. 

Thanks,
    Ohad. 

PS
   I thought I could start with a title: Genomic Range Engineer. Moving up, how about, "Senior range engineer", "Staff range engineer", and finally, "SNP engineer for half open intervals".

mlin commented on src/BCFKeyValueData.cc:
```diff
@@ -174,30 +174,58 @@ Status BCFKeyValueData::dataset_bcf_header(const string& dataset,
+        std::to_string(rng.beg) + ":" +
+        std::to_string(rng.end) + ":" +
+        dataset;
+}
```
It would be nice to ensure the buckets are actually stored totally ordered by genomic range. This won't be the case here because the keys are sorted lexicographically, so for example 21:300000:3300000 precedes 21:90000:120000. This is not a dealbreaker though, feel free to file as a future todo.

mlin commented on src/BCFKeyValueData.cc:
```diff
@@ -292,30 +338,40 @@ Status BCFKeyValueData::import_gvcf(const DataCache* cache,
+        // Make sure a record is not longer than [GENOMIC_INTERVAL_LEN].
+        // If this is not true, then we need to compensate in the search
+        // routine.
+        assert(vt->rlen <= GENOMIC_INTERVAL_LEN);
```
Instead of `assert` let's make this case return Status::Invalid with an informative error message.

mlin commented on src/BCFKeyValueData.cc:
```diff
@@ -217,6 +246,51 @@ Status BCFKeyValueData::dataset_bcf(const string& dataset,
+// Return value:
+// - If none of buckets exist, return NotFound.
+// - If at least one of the buckets exist, but no records were found, return an empty list.
+// - Otherwise, return the list of records that overlap with the query range.
```
We should return Status::NotFound only in some erroneous condition. Zero results for a query is not an error. Exhibiting different behavior based on whether certain internal buckets exist or not is an abstraction violation.

mlin commented on src/BCFSerialize.cc:
```diff
@@ -125,7 +125,6 @@ BCFWriter::~BCFWriter() {
 }
 
 Status BCFWriter::write(bcf1_t* x) {
-    num_entries_ += 1;
```
Did you mean to delete this?

orodeh replied:
Yes, it was a duplicate.

mlin commented on test/BCFKeyValueData.cc:
```diff
@@ -329,8 +329,7 @@ TEST_CASE("BCFKeyValueData BCF retrieval") {
         s = data->dataset_bcf("NA12878D", hdr.get(), range(0, 0, 1000), records);
-        REQUIRE(s.ok());
-        REQUIRE(records.size() == 0);
+        REQUIRE((s == StatusCode::NOT_FOUND || records.size() == 0));
```
Isn't this test deterministic? Why are we asserting two possible behaviors?

orodeh replied:
It is not deterministic, based on bucket size choices. We might have empty buckets, due to the different choices of interval lengths.

mlin commented on src/BCFKeyValueData.cc:
```diff
@@ -292,30 +338,40 @@ Status BCFKeyValueData::import_gvcf(const DataCache* cache,
+        S(write_bucket(body_->db, writer.get(), dataset, bucket));
     }
     writer = NULL;  // clear the memory state
 
```
It would be nice to provide atomic commit of all the key-value pairs for a gVCF file. (Again feel free to file as a future todo)

orodeh replied:
Are you sure it is possible to do this for arbitrary sized data?

mlin commented on src/BCFKeyValueData.cc:
```diff
@@ -292,30 +338,40 @@ Status BCFKeyValueData::import_gvcf(const DataCache* cache,
+                S(write_bucket(body_->db, writer.get(), dataset, bucket));
 
             // start a new in-memory chunk
             writer = NULL;
```
style nit, `nullptr` replaces `NULL` in C++14

mlin commented:
Two overall:

- See if you can factor all the logic about defining the bucket ranges into a separate helper class which `BCFKeyValueData` can ask, "which buckets should I scan for this query range?" and "which bucket does a `bcf1_t` with this range go into?". This will make it easier to change how we define the buckets later.
- We need some specific tests of the bucket logic- e.g. with records and query ranges that span bucket boundaries.

orodeh commented:
Spent some time addressing these comments. I hope the code is better now, please take a look.

mlin commented:
Great! Comments:

* The entire bucket concept should be abstracted inside `BCFKeyValueData`. The caller of `BCFKeyValueData` should not be exposed to it at all. Two leaks currently:
  * `BCFKeyValueData::dataset_bcf` should not behave differently based on the existence or non-existence of buckets. For a valid query with no results, I think it should return an empty vector, returning an error only if something is terribly wrong (e.g. invalid `rid`, negative positions)
  * `BucketExtent` and `BCFBucketRange` don't need to be "exported" in the header. (In another language, these would be internals in the `.ml` and not part of the `.mli` module interface)
* How about running the tests with a couple different bucket sizes, perhaps the default, and then 9?
* There are a couple remaining occurrences of `NULL` :)
* Re: atomic commit, the data set size is not truly unbounded, at least if we restrict ourselves to single-sample gVCFs. They will top out at a couple of gigabytes uncompressed. Anyway, let's come back to this later.

mlin commented:
I also noticed the coverage report suggests that BCFBucketRange::bucket() is not used? https://coveralls.io/builds/3517281/source?filename=src%2FBCFKeyValueData.cc

orodeh commented:
Fixed these issues:
1) Moved the helper classes into the BCFKeyValue.cc file, and removed them from from the header.
2) Replaced NULL with nullptr
3) Made the dataset_bcf function simpler, by returning NotFound only in extreme cases. One case that is problematic, is checking if an rid is valid. To do this, you need the DataCache, because it holds the contigs. 
4) I noticed that I haven't

orodeh commented:
Added tests for various buckets sizes, and fixed the bugs that emerged.

orodeh commented:
This branch should now be ready.

mlin commented on src/BCFKeyValueData.cc:
```diff
@@ -216,6 +380,46 @@ Status BCFKeyValueData::dataset_bcf(const string& dataset,
+
+    // basic sanity checks
+    if (query.rid < 0 || query.beg < 0 || query.end < 0)
+        return Status::NotFound();
```
Let's make this `Status::Invalid("BCFKeyValueData::dataset_bcf: invalid query range", query.str())`. We should make error messages uniquely identifiable whenever possible since we won't have stack traces.  LGTM to squash and merge otherwise. Thanks!
